Handle Firefox Account creation. (#30)

cosmin-malutan opened the pull request:
This pr is for issue #30, please keep in mind that is WIP pr and I have to rebase it after pr #32 get's landed.

Notes:
I packaged the fxa-python-client and I install it with `pip install fxa-python-client-1.0.0.tar.gz`, so I don't have to do other hacks (to move over the setup.py location and to return, to install git on slaves or to clone the workspace from another job and to exlude the trigger.py when removing the repository)
The tradeoff of this was that I had to make a requirenmens text file and install them with pip install -r

Because we have to add the credentials in config file before ve have an environment, I declare the at the top of the script and create the account just before I ran the testrun. If an exception is thrown and we fail to create the account the build will be aborted, so this is safe. 

For a uid at emai creation I made use of BUILD_TAG variable because is the only one is unique, the BUILD_ID is the timestamp and when running concurrent builds it might duplicate.
https://wiki.jenkins-ci.org/display/JENKINS/Building+a+software+project#Buildingasoftwareproject-below

cosmin-malutan committed:
Handle Firefox Account creation. (#30)

whimboo commented on jenkins-master/jobs/mozilla-aurora_fx-account/config.xml:
```diff
@@ -54,14 +54,14 @@
     </hudson.plugins.copyartifact.CopyArtifact>
     <hudson.plugins.xshell.XShellBuilder plugin="xshell@0.9">
-      <commandLine>python trigger.py --urltest $URL_TESTS --urlbuild $URL_BUILD --debug $DEBUG</commandLine>
+      <commandLine>python trigger.py --urltest $URL_TESTS --urlbuild $URL_BUILD --debug $DEBUG --uid $BUILD_TAG</commandLine>
```
I don't see why we have to pass in a UID here. It can be created by the trigger script itself.

whimboo replied:
See https://docs.python.org/2/library/uuid.html

whimboo commented on jenkins-master/jobs/tools/workspace/fxa-python-client-requirenments.txt:
```diff
@@ -0,0 +1,7 @@
+PyBrowserID
+PyHawk
+scrypt
+six
```
All those are in setup.py, so I don't see why we need requirements.txt file.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,139 @@
+
+    # We declare the credentials here and we create the account
+    # after we install the fxa-python-client
+    password = 'mozauto'
```
Please use a random string, so really no-one could interfere here.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,139 @@
+    # We declare the credentials here and we create the account
+    # after we install the fxa-python-client
+    password = 'mozauto'
+    username = 'TPS_%s@restmail.net' % options.uid
```
Please declare variables when needed and not dozen of lines before.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,139 @@
+        subprocess.check_call(['pip', 'install', '-q', '-r',
+                               'fxa-python-client-requirenments.txt'])
+        subprocess.check_call(['pip', 'install',
+                               'fxa-python-client-1.0.0.tar.gz'])
```
The first command should totally not be necessary. Just install the package you created and it should be fine. It's setup.py script already contains everything, as noted above.

cosmin-malutan replied:
If we want to install it via `pip install fxa-python-client-1.0.0.tar.gz` we have to install the requirements separately. it wont install them by it's own. If we want them to be installed by the setup.py, we have to clone the repo change the directory, call `python setup.py install` but this is even more ugly. as I mentioned in notes. I searched about this and I've got feedback from Bebe fro webqa. 
As I sayd the alternative would be to create a job to clone the github repo clone the workspace and call `setup.py install`

whimboo replied:
What? If I have a locally created Mozmill sdist package, and install it, also all dependencies are getting installed. Not sure why this should not work in this case.

cosmin-malutan replied:
I just tested this again, I created an env, called `pip install dist/fxa-python-client-1.0.0.tar.gz` and then called fxa-client, and I've got import error module.

whimboo replied:
All fine here:

> Successfully installed cryptography requests PyBrowserID PyHawk scrypt six fxa-python-client cffi pycparser

You should tell how you created the archive. Also have you opened a PR for the version change? I don't want to see that included as long as it has not been landed. Until then we should leave it without the change.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,139 @@
+                               '-p', password,
+                               'verify'])
+    except subprocess.CalledProcessError as e:
+        sys.exit(e)
```
You can put this in try/except if you want specific error messages. Otherwise it's not necessary. Also keep in mind that if verify fails, we want to remove the account.

whimboo replied:
Also all this code can be moved down.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,139 @@
+        subprocess.check_call(['fxa-client',
+                               '-e', username,
+                               '-p', password,
+                               'destroy'])
```
This will never destroy the created account if test failures happen, because this code is never reached.

cosmin-malutan commented:
@whimboo I added some updates, tomorrow I'm in PTO, I will try to get online to update the pr if needed.

whimboo commented:
fxa-python-client is alive on pypi now as version 1.0.0: https://pypi.python.org/pypi/fxa-python-client

So lets grab it from there now.

cosmin-malutan commented:
@whimboo I made use of pypi, also I made the teardown function remove_account to be called at the end of the script, if testrun get's aborted or if verify  process gets aborted. 
The username and password has to be global variables so we can access the in remove method.
On fresh instances the `pip install fxa-python-client` get's aborted with:

```
19:27:08 
19:27:08 OS/X: confusion between 'cc' versus 'gcc' (see issue 123)
19:27:08 
19:27:08 will not use '__thread' in the C code
19:27:08 
19:27:08 clang: error: unknown argument: '-mno-fused-madd' [-Wunused-command-line-argument-hard-error-in-future]
19:27:08 
19:27:08 clang: note: this will be a hard error (cannot be downgraded to a warning) in the future
19:27:08 
19:27:08 Traceback (most recent call last):
19:27:08 
19:27:08   File "<string>", line 17, in <module>
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/build/cryptography/setup.py", line 174, in <module>
19:27:08 
19:27:08     "test": PyTest,
19:27:08 
19:27:08   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/core.py", line 112, in setup
19:27:08 
19:27:08     _setup_distribution = dist = klass(attrs)
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 260, in __init__
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 284, in fetch_build_eggs
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 563, in resolve
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 799, in best_match
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 811, in obtain
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/dist.py", line 327, in fetch_build_egg
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 446, in easy_install
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 476, in install_item
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 655, in install_eggs
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 930, in build_and_install
19:27:08 
19:27:08   File "/Users/svuser/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/setuptools/command/easy_install.py", line 921, in run_setup
19:27:08 
19:27:08 distutils.errors.DistutilsError: Setup script exited with error: command 'cc' failed with exit status 1
19:27:08 
19:27:08 ----------------------------------------
19:27:08 Cleaning up...
19:27:08 Command python setup.py egg_info failed with error code 1 in /Users/svuser/workspace/mozilla-central_fx-account/tps-env/build/cryptography
```

On both an old and new windows instance I get:

```
19:54:40 Downloading: firefox-33.0a1.en-US.win64-x86_64.tests.zip
19:54:40 Traceback (most recent call last):
19:54:40   File "trigger.py", line 148, in <module>
19:54:40     main()
19:54:40   File "trigger.py", line 105, in main
19:54:40     'create'])
19:54:40   File "C:\Python27\lib\subprocess.py", line 535, in check_call
19:54:40     retcode = call(*popenargs, **kwargs)
19:54:40   File "C:\Python27\lib\subprocess.py", line 522, in call
19:54:40     return Popen(*popenargs, **kwargs).wait()
19:54:40   File "C:\Python27\lib\subprocess.py", line 710, in __init__
19:54:40     errread, errwrite)
19:54:40   File "C:\Python27\lib\subprocess.py", line 958, in _execute_child
19:54:40     startupinfo)
19:54:40 WindowsError: [Error 2] The system cannot find the file specified
```

This is most likely because there is no executable for this under env/Scripts directory.
This happens when we install the fxa-python-client via pip, it says fxa-client is not a recognisable program.

cosmin-malutan commented:
The issue I mentioned on linux and mac does not reproduce on nodes I use daily and I guess I have a newer version of gcc, I didn't had time to check that.
Regarding the issue on windows I don't remember to saw that when I used `pip install dis.zip` but I don't know for sure, I'll check that tomorrow morning.

whimboo commented on jenkins-master/jobs/mozilla-central_fx-account/config.xml:
```diff
@@ -54,14 +54,14 @@
       <doNotFingerprintArtifacts>false</doNotFingerprintArtifacts>
     </hudson.plugins.copyartifact.CopyArtifact>
     <hudson.plugins.xshell.XShellBuilder plugin="xshell@0.9">
-      <commandLine>python trigger.py --urltest $URL_TESTS --urlbuild $URL_BUILD --debug $DEBUG</commandLine>
```
Why this removal? It's totally unrelated to this patch.

whimboo commented on jenkins-master/jobs/release-mozilla-beta_fx-account/config.xml:
```diff
@@ -54,14 +54,14 @@
     </hudson.plugins.copyartifact.CopyArtifact>
     <hudson.plugins.xshell.XShellBuilder plugin="xshell@0.9">
-      <commandLine>python trigger.py --urltest $URL_TESTS --urlbuild $URL_BUILD --debug $DEBUG</commandLine>
+      <commandLine>python trigger.py --urltest $URL_TESTS --urlbuild $URL_BUILD --debug $DEBUG --uid $BUILD_TAG</commandLine>
```
Why thas this been left in here? Please check your patches before uploading them to github.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -70,15 +66,22 @@ def main():
+    # We declare the credentials here and we create the account
+    # after we install the fxa-python-client
+    global password
+    global username
```
I dislike global variables declared this way. Make them real global, or better lets move back to a class, so they can be stored as class members.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -70,15 +66,22 @@ def main():
+    # after we install the fxa-python-client
+    global password
+    global username
+    password = 'crossweaveservicescrossweaveservices'
```
You haven't accounted my comment in the last review for this line. Make it also dynamic.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -70,15 +66,22 @@ def main():
+    global password
+    global username
+    password = 'crossweaveservicescrossweaveservices'
+    username = 'TPS_%s@restmail.net' % uuid.uuid1()
```
Usually you are right in sorting by alphabet, but seeing username coming after password is a bit scary. Lets exchange those lines.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -93,6 +96,26 @@ def main():
+                               '-p', password,
+                               'create'])
+    except subprocess.CalledProcessError as e:
+        sys.exit(e)
```
I don't see why we have to wrap those lines in try/except. It's better you refactor this file a bit for better methods, which handle setup, the execution, and teardown. Call those appropriately from the main method, and handle exceptions and specifically programm exits there.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -93,6 +96,26 @@ def main():
+                               '-p', password,
+                               'verify'])
+    except subprocess.CalledProcessError as e:
+        remove_account()
```
Same as above. Please create more modular code, and don't try to put everything into a big and massive method.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -105,7 +128,21 @@ def main():
+        remove_account()
         sys.exit(e)
 
+    remove_account()
```
This will not be called when any other exception is raised, and the account will remain. Also I think we should print the username/password to the console. So in cases this happens, we could manually delete the account.

cosmin-malutan commented:
I updated the pr but I still have an issue on windows because the fxa-client executable is not found under env/Scripts directory.
As Henrik suggested on irc I should load the python module inside the script and call it from there. To do that I will have to duplicate the code from https://github.com/mozilla/fxa-python-client/blob/master/bin/fxa-client as it can't be imported as a module.
If whatsoever we want to load it as s ingle file using imp.load_source we have to get the path manually, and after we have the exact file path would be easier to just call it as a python script as we do fro 'activate_this.py'

@whimboo what do you think, should I get the path and call the fxa-client script as a subprocess?

whimboo commented:
The problem is that the package does not specify any entry point as you can see here for mozmill:
https://github.com/whimboo/mozmill/blob/master/mozmill/setup.py#L48

Once that is done, a script will be placed in bin or Scripts on Windows. Maybe some refactoring of fxa-client is necessary here. Shall I work on that?

cosmin-malutan commented:
I made the pr https://github.com/mozilla/fxa-python-client/pull/11 for this.
Also  notice, for fxa-python-client we have to install:
~OpenSSL on windows
~libffi-dev and libssl-dev on linux

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+        # We declare the credentials here and we create the account
+        # after we install the fxa-python-client
+        self.username = 'TPS_%s@restmail.net' % uuid.uuid1()
+        self.password = 'TPS_%s' % uuid.uuid1()
```
Please see how the tool is creating a random username and password: https://github.com/mozilla/fxa-python-client/blob/master/bin/fxa-client#L187. This should be enough as using uuid here.

whimboo replied:
Also please keep the username and password as lowercase letters.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+        self.username = 'TPS_%s@restmail.net' % uuid.uuid1()
+        self.password = 'TPS_%s' % uuid.uuid1()
+        print 'TPS username= %s' % self.username
+        print 'TPS password= %s' % self.password
```
Have you tested this code and checked the output? You would see that this output should be visible twice now, given that fxa-client also outputs it: https://github.com/mozilla/fxa-python-client/blob/master/bin/fxa-client#L27

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+                      help='The URL to the Firefox tests on the FTP server')
+    (options, args) = parser.parse_args()
+
+    debug = True if (options.debug == 'true') else False
```
Can't we simply set the default to `None` and pass in `bool(options.debug)` as parameter?

cosmin-malutan replied:
No, because we always get this parameter from the job command and it will never be none, this is the workaround for option arguments for job command.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+
+        # Download the build and tests package
+        self.build = self.download(url=build, destination='builds')
+        self.packaged_tests = self.download(url=tests)
```
`__init__` should really only be used to initialize the class. You are still munging everything into this method.

andreieftimie replied:
I agree, we should have a method that downloads and unzips all assets, and that should be called in the `try` block after we have initialized the module.

whimboo commented:
> Also notice, for fxa-python-client we have to install:
> ~OpenSSL on windows
> ~libffi-dev and libssl-dev on linux

Would be good to file this as a separate issue because it has to be done before we can land this PR. Please give as many as possible details and links when filing the issue.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+        with zipfile.ZipFile(self.packaged_tests) as zip:
+            zip.extractall('tests')
+
+        self.create_environment()
```
I would call it `setup_env()`, so that it also includes any kind of installation of additional tools.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+        # If verifying the account fails, we remove the account
+        self.execute_command(arguments=['fxa-client', '-e', self.username,
+                                        '-p', self.password, 'verify'],
+                             teardown_callback=self.remove_account)
```
Ok, so that one has to be changed, once the dependency on the script issue has been fixed for fxa-client.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+        except subprocess.CalledProcessError as e:
+            if teardown_callback:
+                teardown_callback()
+            sys.exit(e)
```
I don't see a reason why we need this wrapper method. Also please read again my last review, where I mentioned to remove the try/catch. This is not doing what you expect!

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+        self.execute_command(arguments=tps_cmd_args,
+                              teardown_callback=self.remove_account)
+
+        self.remove_account()
```
This will never be called if e.g. mozinstall fails, so the Firefox Account will remain. You should really better try/catch in `main()`, and add a `run()` method to the TPS class.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+                teardown_callback()
+            sys.exit(e)
+
+    def remove_account(self):
```
This method should be called `cleanup()`. You won't know if more has to be added.

cosmin-malutan commented:
I made the changes, required and with the new version of fxa-python-client we don't have the issue anymore on windows, but a new issue I can see on all platforms now is that it installs an dependency but it doesn't find it. for example:

```
15:22:51 Successfully installed argparse cryptography requests PyBrowserID PyHawk six fxa-python-client cffi pycparser
15:22:51 Cleaning up...
15:22:51 Traceback (most recent call last):
15:22:51   File "/Users/danielapetrovici/workspace/mozilla-central_fx-account/tps-env/bin/fxa-client", line 5, in <module>
15:22:51     from pkg_resources import load_entry_point
15:22:51   File "/Users/danielapetrovici/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 2603, in <module>
15:22:51   File "/Users/danielapetrovici/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 666, in require
15:22:51   File "/Users/danielapetrovici/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 565, in resolve
15:22:51 pkg_resources.DistributionNotFound: six
15:22:51 Traceback (most recent call last):
15:22:51   File "/Users/danielapetrovici/workspace/mozilla-central_fx-account/tps-env/bin/fxa-client", line 5, in <module>
15:22:51     from pkg_resources import load_entry_point
15:22:51   File "/Users/danielapetrovici/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 2603, in <module>
15:22:51   File "/Users/danielapetrovici/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 666, in require
15:22:51   File "/Users/danielapetrovici/workspace/mozilla-central_fx-account/tps-env/lib/python2.7/site-packages/setuptools-0.6c11-py2.7.egg/pkg_resources.py", line 565, in resolve
15:22:51 pkg_resources.DistributionNotFound: six
15:22:51 Downloading: builds/firefox-33.0a1.en-US.mac.dmg
15:22:51 Downloading: firefox-33.0a1.en-US.mac.tests.zip
15:22:51 Traceback (most recent call last):
15:22:51   File "trigger.py", line 138, in <module>
15:22:51     main()
15:22:51   File "trigger.py", line 132, in main
15:22:51     run_tps.cleanup()
15:22:51   File "trigger.py", line 36, in cleanup
15:22:51     '-p', self.password, 'destroy'])
15:22:51   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 504, in check_call
15:22:51     raise CalledProcessError(retcode, cmd)
15:22:51 subprocess.CalledProcessError: Command '['fxa-client', '-e', 'fxa-22572df4377f@restmail.net', '-p', 'a802', 'destroy']' returned non-zero exit status 1
```

cosmin-malutan commented:
The issue here was https://github.com/mozilla/fxa-python-client/issues/12, and it will be fixed in mozilla-central I guess, by switching to a newer version of virtualenv https://github.com/mozilla/gecko-dev/blob/master/testing/tps/create_venv.py#L36 
With those in mind this pr is ready for review, @whimboo if you have time can you have a look.

whimboo commented:
This PR is totally not ready for review. @cosmin-malutan I mentioned it multiple times that you should really check the PR on github first, before asking for review. While @andreeamatei is out, I want that @andreieftimie is checking what you are doing here. I hope this will work.

andreieftimie commented:
Sure, I'll review this PR.
But I know very little of the TPS CI env, so not sure how valuable my input there might be.

andreieftimie commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,137 @@
+
+    def download(self, url, destination=None):
+        if destination:
+            if not os.path.exists(destination):
```
You should probably first do a `destination = os.path.abspath(destination)`

andreieftimie commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,137 @@
+                os.makedirs(destination)
+            file_name = os.path.join(destination, url.split('/')[-1])
+        else:
+            file_name = url.split('/')[-1]
```
Actually why not change this whole block to be more pythonic. Try first, ask permission later:

```
try:
    os.makedirs(destination)
except:
    pass
file_name = os.path.join(destination, url.split('/')[-1])
```

Just mark the default `destination` value to be `''` and you're set.

andreieftimie commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,137 @@
+            env_activate_file = os.path.join(tps_env, 'Scripts',
+                                             'activate_this.py')
+        else:
+            env_activate_file = os.path.join(tps_env, 'bin', 'activate_this.py')
```
Can we change this whole block to something like:

```
dir = 'Scripts' if sys.platform == 'win32' else 'bin'
env_activate_file = os.path.join(tps_env, dir, 'activate_this.py')
```

andreieftimie commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,137 @@
+
+        # Install the Firefox and get the binary
+        build_path = mozinstall.install(self.build,
+                                        os.path.join(os.getcwd(), 'binary'))
```
We already have cwd as `current_location`

andreieftimie commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,137 @@
+
+        # We declare the credentials here and we create the account
+        # after we install the fxa-python-client
+        self.username = "fxa-%s@restmail.net" % os.urandom(6).encode("hex")
```
nit: use single quotes for consistency

andreieftimie commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,137 @@
+        # We declare the credentials here and we create the account
+        # after we install the fxa-python-client
+        self.username = "fxa-%s@restmail.net" % os.urandom(6).encode("hex")
+        self.password = os.urandom(2).encode("hex")
```
Why not the same length as the username?

cosmin-malutan commented:
@whimboo I made the updates Andrei requested, also I checked it with pep8 and I added 2 empty lines.
This still needs https://github.com/mozilla/fxa-python-client/issues/12 fixed.

whimboo commented:
Cosmin, you didn't check the pushed changes again. This PR can't still be merged.

cosmin-malutan commented:
@whimboo if you're referring to rebase you mentioned a couple of times that I should do that once the review is complete.

whimboo commented:
This is about merge conflicts, but nothing about a rebase. PRs have to be created based on actual code and not old and mostly outdated one. Totally the same for all of our other repositories. So I don't see what you think differently here.

cosmin-malutan commented:
@whimboo that's why we have merge conflicts, because I didn't do rebase based on jenkins branch. Now that I did that, I had to force push the pr branch, and you explicitly told me that I shouldn't do that until the review is done.

whimboo commented:
This is absolutely not true. Not sure where you get those statements from!? I would advice you to really dig into the git documentation about the `merge` command. I totally do not understand, why you used `rebase` here when we explicitly told you not to do so. You could have asked about the process, but this did not happen, once again. But thankfully you haven't squashed/fixed-up all the commits. So that's a good thing.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+
+        # We declare the credentials here and we create the account
+        # after we install the fxa-python-client
+        self.username = 'fxa-%s@restmail.net' % os.urandom(6).encode('hex')
```
Both of those lines were examples. What we want here is a prefix which indicates that those accounts are used by tps-ci. Right now this tells nothing about it.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+            os.makedirs(destination)
+        except:
+            pass
+        file_name = os.path.join(destination, url.split('/')[-1])
```
So how do you know if the destination folder exists now? It could also be that insufficient permissions stop us here, and we wont recognize it. So silently ignoring all exception cases makes me worried.

cosmin-malutan replied:
If we have a destination argument and we fail to create directory we will fail when trying to write down the file.
IOError: [Errno 2] No such file or directory:

whimboo replied:
You are hiding the real error here. This makes it very hard to investigate. In general plain except should be avoided.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+        # After activating the environment we import the mozinstall module
+        import mozinstall
+
+        # Install the Firefox and get the binary
```
nit: Please remove `the` in-front of Firefox

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+
+    def setup_env(self):
+        # Download the build and tests package
+        self.build = self.download(url=self.build_url, destination='builds')
```
Given that we download the tests into the workspace directly, we should also do this for the build. Not sure why we use `builds` here.

cosmin-malutan replied:
Then should I remove this and the whole directory creation code?

whimboo replied:
That may be easier then.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+        self.binary = mozinstall.get_binary(build_path, 'firefox')
+
+        # Install the fxa-python-client
+        subprocess.check_call(['pip', 'install', 'fxa-python-client'])
```
Keep all the setup steps together. I don't see why this has to happen after Firefox got installed.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,143 @@
+    def cleanup(self):
+        # Remove the account
+        subprocess.check_call(['fxa-client', '-e', self.username,
+                               '-p', self.password, 'destroy'])
```
Where do we uninstall Firefox? Make sure you clean-up everything correctly.

cosmin-malutan commented:
I never before merged the master back in to the issue branch to fix conflicts, or either saw anyone here doing so. I said in comment https://github.com/mozilla/coversheet/pull/34#issuecomment-48296488 that I known that I had conflicts and I expected to fix them at the end, nothing from your second comment told me how should I proceed here, I very well could understand that I should make a new branch based on the newest code and add my changes. I did a rebase on target branch as this is the way to fix merge conflicts in a pr. If we have specific procedure here I really don't know anything about it, I really would like to about it.
I also consulted with the team and considered this is the way to proceed.

whimboo commented:
> I never before merged the master back in to the issue branch to fix conflicts, or either saw anyone here doing so.

I don't see why you want to merge in anything from master. This will certainly fail. What you need are the latest changes from the jenkins branch.

> I said in comment #34 (comment) that I known that I had conflicts and I expected to fix them at the end, nothing from your second comment told me how should I proceed here, I very well could understand that I should make a new branch based on the newest code and add my changes. I did a rebase on target branch as this is the way to fix merge conflicts in a pr.

If things are unclear, and well I also can miss some comments, you should communicate on IRC. I haven't seen anything like that today when you and @andreieftimie discussed that problem. Anyway, it's off-topic and if necessary needs a separate discussion. Otherwise see my review comments I gave about 2h ago.

cosmin-malutan commented:
I updated the patch due to review:
I uninstall Firefox in cleanup
I added the configuration for concurrent builds
I removed the destination argument in download method
I ran the jobs on all platforms, it works on mac and linux but it fails on windows with tha latest tinderbox:
http://pastebin.mozilla.org/5533850

@whimboo can you review this please?

cosmin-malutan commented:
It works on windows too, it seems that I started the slave-agent client form an environment and that causesed that exception. So all is good

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,141 @@
+        # We declare the credentials here and we create the account
+        # after we install the fxa-python-client
+        self.username = 'tps_ci-%s@restmail.net' % os.urandom(6).encode('hex')
+        self.password = 'tps_ci-%s' % os.urandom(2).encode('hex')
```
The change for the password was not necessary. If you don't mind please revert it.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,141 @@
+                               '-p', self.password, 'destroy'])
+
+        import mozinstall
+        mozinstall.uninstall(self.build_path)
```
Keep in mind that if the fxa-client call fails, the Firefox instance will not be uninstalled!

cosmin-malutan commented:
@whimboo I updated the pr, if you want to implement the usage of preconfigured environments in this pr I think it would be best to talk about how we would do that, and how this is gonna affect this pr.

whimboo commented:
No, the pre-configured environment is something for later. Right now we need this feature being present.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,149 @@
+            mozinstall.uninstall(self.build_path)
+        except:
+            print "Failed to remove the Firefox build, " \
+                  "please remove it manually"
```
Please keep in mind that even when cleanup fails, the script has to exit with an exit code not 0. This won't happen here.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,149 @@
+        run_tps.cleanup()
+        sys.exit(e)
+
+    run_tps.cleanup()
```
Maybe refactor this a bit, so that we call cleanup() in `finally`.

cosmin-malutan commented:
@whimboo I made the changes, also if we are not gonna make use of preconfigured environments we would have to install the dependencies on client nodes as bug https://bugzilla.mozilla.org/show_bug.cgi?id=1036321 is closed.
On windows the process is a bit more difficult, so I want to add the steps somewhere to don't forget them.

whimboo commented:
So lets get them added. We can have a wonderful readme for that repository.

cosmin-malutan commented:
In this pr or on another one as followup?

whimboo commented:
Well, you are introducing fxa-python-client here, so why should it not be included in this PR? People who wanna use our CI won't be able to run it because of missing dependencies.

whimboo commented on jenkins-master/jobs/tools/workspace/trigger.py:
```diff
@@ -0,0 +1,154 @@
+
+        # We declare the credentials here and we create the account
+        # after we install the fxa-python-client
+        self.username = 'tps_ci-%s@restmail.net' % os.urandom(6).encode('hex')
```
I missed that in my last reviews. The tool is known as `coversheet` and not `tps_ci`. So please name it accordingly.

whimboo commented:
> Well, you are introducing fxa-python-client here, so why should it not be included in this PR? 

Maybe you could create a simple entry in the readme file which points to the issue which is about the documentation part. I think that would also help and we can get this landed ASAP.

whimboo commented:
Otherwise this PR looks fine and we should get it landed. Please disable OS X and Windows in the config files for pulse for now. We have to make linux stable first.

cosmin-malutan commented:
@whimboo I disabled the jobs on mac and windows and I tested it with `--push-message` argument. I updated the prefix for username. Regarding the platform requirements I filed issue #37

whimboo commented:
I still don't see that the README file has been updated with the link to the newly filed issue for updating the requirements. Please do that, and squash all the commits.

cosmin-malutan commented:
@whimboo  Done!
